plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -648,13 +648,18 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
             const auto block_deadline = calculate_block_deadline( chain.pending_block_time() );
             push_result pr = push_transaction( block_deadline, trx, persist_until_expired, return_failure_trace, next );
 
-            exhausted = pr.block_exhausted;
             if( pr.trx_exhausted ) {
                _unapplied_transactions.add_incoming( trx, persist_until_expired, return_failure_trace, next );
             } else if( pr.persist ) {
                _unapplied_transactions.add_persisted( trx );
             }
 
+            exhausted = pr.block_exhausted; // transaction exhausted if block was exhausted
```

Comment:
I think remove the comment. `exhausted` here means block is exhausted. That is why the comment below is needed. We say block is exhausted when trx is exhausted below because we want to restart the block when in speculative mode.